src/include/migraphx/op/nonmaxsuppression.hpp

```diff
@@ -265,31 +267,33 @@ struct nonmaxsuppression
             auto batch_boxes_start = boxes.begin() + batch_idx * num_boxes * 4;
             auto boxes_heap = filter_boxes_by_score(scores_start, num_boxes, score_threshold);
             selected_boxes_inside_class.clear();
-            // Get the next box with top score, filter by iou_threshold
             while(not boxes_heap.empty() &&
                   selected_boxes_inside_class.size() < max_output_boxes_per_class)
             {
-                // Check with existing selected boxes for this class, remove box if it
-                // exceeds the IOU (Intersection Over Union) threshold
+                // select next top scorer box and remove any boxes from boxes_heap that exceeds IOU
+                // threshold with the selected box
                 const auto next_top_score = boxes_heap.top();
-                bool not_selected =
-                    std::any_of(selected_boxes_inside_class.begin(),
-                                selected_boxes_inside_class.end(),
-                                [&](auto selected_index) {
-                                    return this->suppress_by_iou(
-                                        batch_box(batch_boxes_start, next_top_score.second),
-                                        batch_box(batch_boxes_start, selected_index.second),
-                                        iou_threshold);
-                                });
-
-                if(not not_selected)
+                boxes_heap.pop();
+                selected_boxes_inside_class.push_back(next_top_score);
+                selected_indices.push_back(batch_idx);
+                selected_indices.push_back(class_idx);
+                selected_indices.push_back(next_top_score.second);
+                auto temp_heap =
+                    boxes_heap; // priority_queue doesn't support iteration, make a copy
+                std::priority_queue<std::pair<double, int64_t>> remainder;
+                while(!temp_heap.empty())
                 {
-                    selected_boxes_inside_class.push_back(next_top_score);
-                    selected_indices.push_back(batch_idx);
-                    selected_indices.push_back(class_idx);
-                    selected_indices.push_back(next_top_score.second);
+                    auto tmp_top_box = temp_heap.top();
+                    if(not this->suppress_by_iou(
+                           batch_box(batch_boxes_start, tmp_top_box.second),
+                           batch_box(batch_boxes_start, next_top_score.second),
+                           iou_threshold))
+                    {
+                        remainder.push(tmp_top_box);
+                    }
+                    temp_heap.pop();
                 }
-                boxes_heap.pop();
+                boxes_heap = remainder;
```

Comment:
Since you assign `remainder` to `boxes_heap` at the end, then it seems like you dont need to make a copy with `temp_heap`.

Reply:
Done.